include/nbl/ext/DepthPyramidGenerator/DepthPyramidGenerator.h

```diff
@@ -0,0 +1,105 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_EXT_DEPTH_PYRAMID_GENERATOR_H_INCLUDED_
+#define _NBL_EXT_DEPTH_PYRAMID_GENERATOR_H_INCLUDED_
+
+#include "nabla.h"
+
+#include "nbl/video/IGPUImageView.h"
+#include "nbl/asset/format/EFormat.h"
+
+using namespace nbl;
+using namespace nbl::core;
+using namespace nbl::asset;
+using namespace nbl::video;
+
+namespace nbl
+{
+namespace ext
+{
+namespace DepthPyramidGenerator
+{
+
+// TODO: test: `E_MIPMAP_GENERATION_OPERATOR::BOTH`, `reundUpToPoTWithPadding = true`
+
+class DepthPyramidGenerator final
+{
+public:
+
+	enum class E_MIPMAP_GENERATION_OPERATOR
+	{
+		MAX, 
+		MIN,
+		BOTH // min goes to r, min to g
+	};
+	
+	enum class E_WORK_GROUP_SIZE
+	{
+		E16x16x1 = 16u,
+		E32x32x1 = 32u
+	};
+
+	struct Config
+	{
+		E_WORK_GROUP_SIZE workGroupSize = E_WORK_GROUP_SIZE::E32x32x1;
+		asset::E_FORMAT outputFormat = asset::E_FORMAT::EF_R32_SFLOAT;
+		E_MIPMAP_GENERATION_OPERATOR op = E_MIPMAP_GENERATION_OPERATOR::MAX;
+		uint32_t lvlLimit = 0u; //no limit when set to 0 (full mip chain)
+		bool roundUpToPoTWithPadding = false;
+	};
+
+	// inputDepthImageView - input texture
+	//outputDepthPyramidMips - array of created mipMaps
+	DepthPyramidGenerator(IVideoDriver* driver, IAssetManager* am, core::smart_refctd_ptr<IGPUImageView> inputDepthImageView,
+		core::smart_refctd_ptr<IGPUImageView>* outputDepthPyramidMips,
+		const Config& config = Config());
+
+	inline uint32_t getMaxMipCntFromImage(const core::smart_refctd_ptr<IGPUImageView>& image, bool roundUpToPoTWithPadding = false)
+	{
+		const VkExtent3D lvl0MipExtent = calcLvl0MipExtent(image->getCreationParameters().image->getCreationParameters().extent, roundUpToPoTWithPadding);
+
+		uint32_t minVal = std::min(lvl0MipExtent.width, lvl0MipExtent.height);
+
+		uint32_t mipLvlCnt = 1u;
+		while (minVal >>= 1u)
+			mipLvlCnt++;
+
+		return mipLvlCnt;
+	}
+
+	void generateMipMaps();
+
+private:
+	inline VkExtent3D calcLvl0MipExtent(const VkExtent3D& sourceImageExtent, bool roundUpToPoTWithPadding)
+	{
+		VkExtent3D lvl0MipExtent;
+
+		lvl0MipExtent.width = core::roundUpToPoT(sourceImageExtent.width);
+		lvl0MipExtent.height = core::roundUpToPoT(sourceImageExtent.height);
+
+		if (!roundUpToPoTWithPadding)
+		{
+			lvl0MipExtent.width >>= 1u;
+			lvl0MipExtent.height >>= 1u;
+		}
+
+		return lvl0MipExtent;
+	}
+
+	void configureMipImages(core::smart_refctd_ptr<IGPUImageView> inputDepthImageView, core::smart_refctd_ptr<IGPUImageView>* outputDepthPyramidMips, const Config& config);
+
+private:
+	IVideoDriver* m_driver;
+
+	core::smart_refctd_ptr<IGPUDescriptorSet> m_ds = nullptr;
```

Comment:
you want a method to precompute the descriptor set, and one to return the descriptor layout, see other extensions